source/common/network/connection_impl.cc

```diff
@@ -34,6 +34,16 @@ Address::InstanceConstSharedPtr getNullLocalAddress(const Address::Instance& add
 }
 } // namespace
 
+int ConnectionImplUtility::createSocket(Address::InstanceConstSharedPtr address,
+                                        Optional<Address::InstanceConstSharedPtr> source_address) {
+  int fd = address->socket(Address::SocketType::Stream);
+  if (fd >= 0 && source_address.valid()) {
+    int rc = source_address.value()->bind(fd);
+    ASSERT(rc >= 0);
```

Comment:
Seems we should have stronger error handling here. If it really should never happen except for some unrecoverable sitaution, `RELEASE_ASSERT`, otherwise `return -1` or exception.